### PR TITLE
Add support for building to the wasm32-wasi target

### PR DESCRIPTION
**Pull request by ereslibre**
This change performs initial adaptations allowing the PHP interpreter
to be compiled for the wasm32-wasi target.

Challenges found:

- `setjmp`/`longjmp` not available in this platform: got a
contribution to emulate it in order to run shutdown
functions. Currently working on implementing fibers by emulating
setjmp/longjmp.

- Processes: WebAssembly has no concept of processes, so many
process-related operations had to be emulated, or are no-ops. `getpid`
and clocks are emulated. Signal handling has a minimal
emulation. `uid, `gid`, and related concepts that do not map well to a
WebAssembly sandbox are also removed.
  - `mmap` implementation is a minimal emulation that allows to
  allocate empty memory or map file contents to it.

- Filesystem: many filesystem operations cannot be mapped to WASI, and
so these operations are no-ops.
  - `S_IFSOCK` matches `S_IFIFO` in current `wasi-libc`; removed the
  `S_IFIFO` case, because `clang` will fail building due to a
  duplicated constant in `switch` statements.
  - File locks are also stubbed and always return success.

- Networking: given the WebAssembly sandbox and capability-based
security, it is not possible to open a socket as of today. Although
the host is able to open a socket and forward a file descriptor,
leaving the accept() syscall to the WebAssembly module, we have
no-oped networking syscalls at this time.

Co-Authored-By: Asen Alexandrov <alexandrov@vmware.com>

### CODE REVIEW

**Comment by ereslibre** [comment]
cc/ @vmware-labs/webassembly-language-runtimes-maintainers

**Review by assambar — APPROVED** [review]
I love this as a first version of the thing. As we discussed offline we should provide more context for the different types of changes and places we needed to change something.

Then maybe a roadmap of what we think should be done next, and asking for opinion on PHP internals on what the community things we should also do, before merging to master.

**Pull request closed without merging**
